# PCOS Health Analytics Platform

## Notebook 2: Data Cleaning

### Objective

Prepare the main PCOS clinical dataset for analysis by identifying and resolving data quality issues, including missing values, inconsistent column names, unnecessary columns, and data formatting.

### Dataset Used

This notebook uses the **main PCOS clinical dataset** (`PCOS_data_without_infertility.xlsx`), which was explored in Notebook 1.

In [1]:
#Importing Libraries
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)

In [2]:
#Loading the datasets
df = pd.read_excel(
    "../data/raw/PCOS_data_without_infertility.xlsx",
    sheet_name="Full_new"
)

df.head()

,Sl. No,Patient File No.,PCOS (Y/N),Age (yrs),Weight (Kg),Height(Cm),BMI,Blood Group,Pulse rate(bpm),RR (breaths/min),Hb(g/dl),Cycle(R/I),Cycle length(days),Marraige Status (Yrs),Pregnant(Y/N),No. of aborptions,I beta-HCG(mIU/mL),II beta-HCG(mIU/mL),FSH(mIU/mL),LH(mIU/mL),FSH/LH,Hip(inch),Waist(inch),Waist:Hip Ratio,TSH (mIU/L),AMH(ng/mL),PRL(ng/mL),Vit D3 (ng/mL),PRG(ng/mL),RBS(mg/dl),Weight gain(Y/N),hair growth(Y/N),Skin darkening (Y/N),Hair loss(Y/N),Pimples(Y/N),Fast food (Y/N),Reg.Exercise(Y/N),BP _Systolic (mmHg),BP _Diastolic (mmHg),Follicle No. (L),Follicle No. (R),Avg. F size (L) (mm),Avg. F size (R) (mm),Endometrium (mm),Unnamed: 44
0,1,1,0,28,44.6,152.0,19.300000,15,78,22,10.48,2,5,7.0,0,0,1.99,1.99,7.95,3.68,2.160326,36,30,0.833333,0.68,2.07,45.16,17.1,0.57,92.0,0,0,0,0,0,1.0,0,110,80,3,3,18.0,18.0,8.5,NaN
1,2,2,0,36,65.0,161.5,24.921163,15,74,20,11.70,2,5,11.0,1,0,60.80,1.99,6.73,1.09,6.174312,38,32,0.842105,3.16,1.53,20.09,61.3,0.97,92.0,0,0,0,0,0,0.0,0,120,70,3,5,15.0,14.0,3.7,NaN
2,3,3,1,33,68.8,165.0,25.270891,11,72,18,11.80,2,5,10.0,1,0,494.08,494.08,5.54,0.88,6.295455,40,36,0.900000,2.54,6.63,10.52,49.7,0.36,84.0,0,0,0,1,1,1.0,0,120,80,13,15,18.0,20.0,10.0,NaN
3,4,4,0,37,65.0,148.0,29.674945,13,72,20,12.00,2,5,4.0,0,0,1.99,1.99,8.06,2.36,3.415254,42,36,0.857143,16.41,1.22,36.90,33.4,0.36,76.0,0,0,0,0,0,0.0,0,120,70,2,2,15.0,14.0,7.5,NaN
4,5,5,0,25,52.0,161.0,20.060954,11,72,18,10.00,2,5,1.0,1,0,801.45,801.45,3.98,0.90,4.422222,37,30,0.810811,3.57,2.26,30.09,43.8,0.38,84.0,0,0,0,1,0,0.0,0,120,80,3,4,16.0,14.0,7.0,NaN


In [3]:
#Creating a Working Copy
df_clean = df.copy()

In [6]:
#Removing empty columns
print(df_clean.shape)
df_clean = df_clean.drop(columns=["Unnamed: 44"])
print(df_clean.shape)

(541, 45)
(541, 44)


### Observation

The column `Unnamed: 44` was removed because it contained almost entirely missing values and did not represent a meaningful clinical variable.

In [8]:
#Cleaning Column Names
df_clean.columns = (
    df_clean.columns
    .str.strip()
    .str.replace(r"\s+", " ", regex=True)
)

df_clean.columns.tolist()

['Sl. No',
 'Patient File No.',
 'PCOS (Y/N)',
 'Age (yrs)',
 'Weight (Kg)',
 'Height(Cm)',
 'BMI',
 'Blood Group',
 'Pulse rate(bpm)',
 'RR (breaths/min)',
 'Hb(g/dl)',
 'Cycle(R/I)',
 'Cycle length(days)',
 'Marraige Status (Yrs)',
 'Pregnant(Y/N)',
 'No. of aborptions',
 'I beta-HCG(mIU/mL)',
 'II beta-HCG(mIU/mL)',
 'FSH(mIU/mL)',
 'LH(mIU/mL)',
 'FSH/LH',
 'Hip(inch)',
 'Waist(inch)',
 'Waist:Hip Ratio',
 'TSH (mIU/L)',
 'AMH(ng/mL)',
 'PRL(ng/mL)',
 'Vit D3 (ng/mL)',
 'PRG(ng/mL)',
 'RBS(mg/dl)',
 'Weight gain(Y/N)',
 'hair growth(Y/N)',
 'Skin darkening (Y/N)',
 'Hair loss(Y/N)',
 'Pimples(Y/N)',
 'Fast food (Y/N)',
 'Reg.Exercise(Y/N)',
 'BP _Systolic (mmHg)',
 'BP _Diastolic (mmHg)',
 'Follicle No. (L)',
 'Follicle No. (R)',
 'Avg. F size (L) (mm)',
 'Avg. F size (R) (mm)',
 'Endometrium (mm)']

### Column names
Column names were standardized by removing leading/trailing whitespace and replacing multiple consecutive spaces with a single space. This improves readability and reduces the risk of referencing errors during analysis.

In [ ]:
#Handling missing values
missing = df_clean.isnull().sum()

missing[missing > 0]

Marraige Status (Yrs)    1
Fast food (Y/N)          1
dtype: int64

In [10]:
df_clean[df_clean["Fast food (Y/N)"].isnull()]
df_clean[df_clean["Fast food (Y/N)"].isnull()]

,Sl. No,Patient File No.,PCOS (Y/N),Age (yrs),Weight (Kg),Height(Cm),BMI,Blood Group,Pulse rate(bpm),RR (breaths/min),Hb(g/dl),Cycle(R/I),Cycle length(days),Marraige Status (Yrs),Pregnant(Y/N),No. of aborptions,I beta-HCG(mIU/mL),II beta-HCG(mIU/mL),FSH(mIU/mL),LH(mIU/mL),FSH/LH,Hip(inch),Waist(inch),Waist:Hip Ratio,TSH (mIU/L),AMH(ng/mL),PRL(ng/mL),Vit D3 (ng/mL),PRG(ng/mL),RBS(mg/dl),Weight gain(Y/N),hair growth(Y/N),Skin darkening (Y/N),Hair loss(Y/N),Pimples(Y/N),Fast food (Y/N),Reg.Exercise(Y/N),BP _Systolic (mmHg),BP _Diastolic (mmHg),Follicle No. (L),Follicle No. (R),Avg. F size (L) (mm),Avg. F size (R) (mm),Endometrium (mm)
156,157,157,0,27,53.2,158.0,21.3,13,72,22,10.5,4,4,5.0,0,0,1.99,1.99,3.09,0.59,5.237,37,32,0.864,2.03,5.27,7.16,25.3,0.36,91.0,0,0,1,1,0,NaN,1,120,70,5,7,11.0,13.0,11.0


In [11]:
df_clean["Fast food (Y/N)"].value_counts()

Fast food (Y/N)
1.0    278
0.0    262
Name: count, dtype: int64

In [13]:
df_clean["Fast food (Y/N)"] = df_clean["Fast food (Y/N)"].fillna(
    df_clean["Fast food (Y/N)"].mode()[0]
)
df_clean["Fast food (Y/N)"].isnull().sum()

np.int64(0)

### Handling Missing Values: Fast food (Y/N)

The `Fast food (Y/N)` column contained one missing value. Since this represents only **0.18%** of the dataset (1 out of 541 records), the missing value was imputed using the mode (`1`, representing "Yes"). This approach preserves all patient records while having a negligible impact on the overall data distribution.

In [14]:
df_clean[df_clean["Marraige Status (Yrs)"].isnull()]

,Sl. No,Patient File No.,PCOS (Y/N),Age (yrs),Weight (Kg),Height(Cm),BMI,Blood Group,Pulse rate(bpm),RR (breaths/min),Hb(g/dl),Cycle(R/I),Cycle length(days),Marraige Status (Yrs),Pregnant(Y/N),No. of aborptions,I beta-HCG(mIU/mL),II beta-HCG(mIU/mL),FSH(mIU/mL),LH(mIU/mL),FSH/LH,Hip(inch),Waist(inch),Waist:Hip Ratio,TSH (mIU/L),AMH(ng/mL),PRL(ng/mL),Vit D3 (ng/mL),PRG(ng/mL),RBS(mg/dl),Weight gain(Y/N),hair growth(Y/N),Skin darkening (Y/N),Hair loss(Y/N),Pimples(Y/N),Fast food (Y/N),Reg.Exercise(Y/N),BP _Systolic (mmHg),BP _Diastolic (mmHg),Follicle No. (L),Follicle No. (R),Avg. F size (L) (mm),Avg. F size (R) (mm),Endometrium (mm)
458,459,459,1,36,66.0,162.0,25.1,15,72,20,11.0,4,7,NaN,0,0,515.33,1.99,1.64,0.17,9.647059,39,37,0.948718,1.86,6.6,25.0,20.8,0.25,93.0,0,0,0,0,0,0.0,0,120,80,14,5,19.0,19.0,8.0


In [15]:
df_clean["Marraige Status (Yrs)"].median()

np.float64(7.0)

In [17]:
df_clean["Marraige Status (Yrs)"] = df_clean["Marraige Status (Yrs)"].fillna(
    df_clean["Marraige Status (Yrs)"].median()
)
df_clean.isnull().sum().sum()

np.int64(0)

 ### Handling Missing Values: Marraige Status (Yrs)

The `Marraige Status (Yrs)` column contained one missing value (approximately 0.18% of the dataset). Since this is a numerical variable with only one missing observation, the value was imputed using the median (7 years). Median imputation preserves the dataset while minimizing the influence of outliers.

In [18]:
df_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 541 entries, 0 to 540
Data columns (total 44 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Sl. No                 541 non-null    int64  
 1   Patient File No.       541 non-null    int64  
 2   PCOS (Y/N)             541 non-null    int64  
 3   Age (yrs)              541 non-null    int64  
 4   Weight (Kg)            541 non-null    float64
 5   Height(Cm)             541 non-null    float64
 6   BMI                    541 non-null    float64
 7   Blood Group            541 non-null    int64  
 8   Pulse rate(bpm)        541 non-null    int64  
 9   RR (breaths/min)       541 non-null    int64  
 10  Hb(g/dl)               541 non-null    float64
 11  Cycle(R/I)             541 non-null    int64  
 12  Cycle length(days)     541 non-null    int64  
 13  Marraige Status (Yrs)  541 non-null    float64
 14  Pregnant(Y/N)          541 non-null    int64  
 15  No. of aborptions

In [20]:
#Finding objects column
df_clean.select_dtypes(include="object").columns
df_clean["II beta-HCG(mIU/mL)"].head(10)
df_clean["II beta-HCG(mIU/mL)"].unique()[:20]

array([1.99, 494.08, 801.45, 100.51, 158.51, 1214.23, 91.55, 749.98,
       218.65, 13, 610.63, 4490.18, 11.24, 15, 768.03, 12, 10, 20, 30.66,
       1455], dtype=object)

In [21]:
pd.to_numeric(df_clean["II beta-HCG(mIU/mL)"], errors="coerce").isna().sum()

np.int64(1)

In [22]:
mask = pd.to_numeric(df_clean["II beta-HCG(mIU/mL)"], errors="coerce").isna()
df_clean.loc[mask, ["Patient File No.", "II beta-HCG(mIU/mL)"]]

,Patient File No.,II beta-HCG(mIU/mL)
123,124,1.99.


In [23]:
df_clean["II beta-HCG(mIU/mL)"] = df_clean["II beta-HCG(mIU/mL)"].replace(
    "1.99.", "1.99"
)
df_clean["II beta-HCG(mIU/mL)"] = pd.to_numeric(
    df_clean["II beta-HCG(mIU/mL)"]
)
df_clean["II beta-HCG(mIU/mL)"].dtype

dtype('float64')

In [24]:
df_clean.select_dtypes(include="object").columns

Index(['AMH(ng/mL)'], dtype='str')

In [25]:
mask = pd.to_numeric(df_clean["AMH(ng/mL)"], errors="coerce").isna()

df_clean.loc[mask, ["Patient File No.", "AMH(ng/mL)"]]

,Patient File No.,AMH(ng/mL)
305,306,a


In [27]:
df_clean["AMH(ng/mL)"] = df_clean["AMH(ng/mL)"].replace("a", np.nan)
df_clean["AMH(ng/mL)"] = pd.to_numeric(df_clean["AMH(ng/mL)"])
df_clean["AMH(ng/mL)"] = pd.to_numeric(df_clean["AMH(ng/mL)"])
df_clean["AMH(ng/mL)"] = df_clean["AMH(ng/mL)"].fillna(
    df_clean["AMH(ng/mL)"].median()
)
df_clean.select_dtypes(include="object").columns
df_clean.isnull().sum()

Sl. No                   0
Patient File No.         0
PCOS (Y/N)               0
Age (yrs)                0
Weight (Kg)              0
Height(Cm)               0
BMI                      0
Blood Group              0
Pulse rate(bpm)          0
RR (breaths/min)         0
Hb(g/dl)                 0
Cycle(R/I)               0
Cycle length(days)       0
Marraige Status (Yrs)    0
Pregnant(Y/N)            0
No. of aborptions        0
I beta-HCG(mIU/mL)       0
II beta-HCG(mIU/mL)      0
FSH(mIU/mL)              0
LH(mIU/mL)               0
FSH/LH                   0
Hip(inch)                0
Waist(inch)              0
Waist:Hip Ratio          0
TSH (mIU/L)              0
AMH(ng/mL)               0
PRL(ng/mL)               0
Vit D3 (ng/mL)           0
PRG(ng/mL)               0
RBS(mg/dl)               0
Weight gain(Y/N)         0
hair growth(Y/N)         0
Skin darkening (Y/N)     0
Hair loss(Y/N)           0
Pimples(Y/N)             0
Fast food (Y/N)          0
Reg.Exercise(Y/N)        0
B

In [28]:
df_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 541 entries, 0 to 540
Data columns (total 44 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Sl. No                 541 non-null    int64  
 1   Patient File No.       541 non-null    int64  
 2   PCOS (Y/N)             541 non-null    int64  
 3   Age (yrs)              541 non-null    int64  
 4   Weight (Kg)            541 non-null    float64
 5   Height(Cm)             541 non-null    float64
 6   BMI                    541 non-null    float64
 7   Blood Group            541 non-null    int64  
 8   Pulse rate(bpm)        541 non-null    int64  
 9   RR (breaths/min)       541 non-null    int64  
 10  Hb(g/dl)               541 non-null    float64
 11  Cycle(R/I)             541 non-null    int64  
 12  Cycle length(days)     541 non-null    int64  
 13  Marraige Status (Yrs)  541 non-null    float64
 14  Pregnant(Y/N)          541 non-null    int64  
 15  No. of aborptions

In [30]:
#Saving cleaned dataset
import os

os.makedirs("../data/processed", exist_ok=True)
df_clean.to_csv("../data/processed/pcos_cleaned.csv", index=False)
os.listdir("../data/processed")

['pcos_cleaned.csv']

# Notebook Summary

## Data Cleaning Summary

The main PCOS clinical dataset was cleaned and standardized to ensure data quality before exploratory analysis and modeling.

### Cleaning Steps Performed

- Removed the empty `Unnamed: 44` column.
- Standardized column names by removing unnecessary whitespace.
- Assessed and handled missing values.
  - Imputed the missing value in `Fast food (Y/N)` using the mode.
  - Imputed the missing value in `Marraige Status (Yrs)` using the median.
- Corrected formatting inconsistencies in `II beta-HCG(mIU/mL)` and converted the column to a numeric data type.
- Identified and corrected an invalid entry (`"a"`) in `AMH(ng/mL)`, replacing it with the median after conversion to a missing value.
- Verified that all variables have appropriate data types.
- Confirmed that the cleaned dataset contains no missing values.

The cleaned dataset has been saved for use in subsequent notebooks covering exploratory data analysis, SQL analysis, Power BI dashboard development, and machine learning.